polytracker/src/taintdag/util.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2022-present, Trail of Bits, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed in accordance with the terms specified in
+ * the LICENSE file found in the root directory of this source tree.
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+
+#include "taintdag/error.h"
+#include "taintdag/util.h"
+
+using namespace taintdag;
+
+namespace util {
+
+Length Length::from_returned_size(ssize_t retval) {
+  if (retval > 0) {
+    // All positive ssize_t can be represented in size_t
+    return Length{static_cast<size_t>(retval)};
+  }
+  return Length{};
+}
+
+Length Length::from_returned_size_count(size_t size, size_t nitems) {
+
+  size_t byte_size;
+
+  // NOTE: using clang builtin. Assuming portability is not an issue as we rely
+  // heavily on LLVM.
+  static_assert(sizeof(size_t) == sizeof(unsigned long),
+                "Implementation requires size_t is same size as unsigned long");
+  if (__builtin_umull_overflow(size, nitems, &byte_size)) {
+    error_exit("Length size ", size, " nitems ", nitems, " overflows");
+  }
+
+  // Nothing read
+  if (byte_size == 0) {
+    return Length{};
+  }
+
+  // Read ok
+  return Length{byte_size};
+}
+
+Length Length::from_returned_string(char const *str) {
+  if (!str) {
+    return Length{};
+  }
+  return Length{strlen(str)};
```

Comment:
Are you suggesting we change the function signature to accept a `std::string_view` instead of `char*`? E.g. into
```c++
Length Length::from_returned_string(std::string_view str) {
  return Length{str.size()};
}
```
If I understand it correctly it is undefined behaviour to construct a `std::string_view` from a nullptr. As this function is intended to be invoked from e.g. `fgets`, which could return a nullptr we need to check for it. So, one option is
```c++
Length Length::from_returned_string(char const *str) {
  if (str) {
    return Length{std::string_view(str).size()};
  } else {
    return Length{};
  }
}
```
Is this more readable? Thoughts?

Reply:
```C
Length Length::from_returned_string(char const *str) {
  if (str) {
    return Length{std::string_view(str).size()};
  }
  return Length{}; 
}
```

I think the above is ideal.